jvm-diagnostics/src/main/java/com/palantir/jvm/diagnostics/jfr/RecordingStreams.java

```diff
@@ -0,0 +1,450 @@
+ * on newer Java runtimes, while still targeting language levels below JDK14 (when the streaming API was
+ * introduced).
+ *
+ * RecordingStreams provides utility methods for creating a JFR recording stream when the runtime allows it.
+ * The returned object will be shaped similarly to `jdk.jfr.consumer.RecordingStream`. When the runtime does not
+ * support this, an empty Optional is returned.
+ *
+ * See also <a href="https://docs.oracle.com/en/java/javase/17/docs/api/jdk.jfr/jdk/jfr/consumer/RecordingStream.html">
+ *     jdk.jfr.consumer.RecordingStream
+ *     </a>
+ */
+public class RecordingStreams {
+    private static final Logger log = LoggerFactory.getLogger(RecordingStreams.class);
+
+    private RecordingStreams() {}
+
+    /**
+     * Creates an event stream for the current JVM (Java Virtual Machine).
+     *
+     * See <a href="https://docs.oracle.com/en/java/javase/17/docs/api/jdk.jfr/jdk/jfr/consumer/RecordingStream.html#%3Cinit%3E()">
+     *     https://docs.oracle.com/en/java/javase/17/docs/api/jdk.jfr/jdk/jfr/consumer/RecordingStream.html#%3Cinit%3E()
+     *     </a>
+     *
+     * @return an {@link Optional} containing an {@link RecordingStreamSupport} which delegates to an underlying
+     * `jdk.jfr.consumer.RecordingStream` if the runtime supports it, or {@link Optional#empty()} otherwise.
+     */
+    public static Optional<RecordingStreamSupport> newRecordingStream() {
+        return isSupported().map(_v -> {
+            try {
+                return new ReflectiveRecordingStreamSupport(null);
+            } catch (Throwable t) {
+                log.warn("JFR recording stream support is not available", t);
+                return null;
+            }
+        });
+    }
+
+    /**
+     * Creates a recording stream using settings from a configuration.
+     *
+     * See <a href="https://docs.oracle.com/en/java/javase/17/docs/api/jdk.jfr/jdk/jfr/consumer/RecordingStream.html#%3Cinit%3E(jdk.jfr.Configuration)">
+     *     https://docs.oracle.com/en/java/javase/17/docs/api/jdk.jfr/jdk/jfr/consumer/RecordingStream.html#%3Cinit%3E(jdk.jfr.Configuration)
+     *     </a>
+     *
+     * @param configuration configuration that contains the settings to use, not null
+     * @return an {@link Optional} containing an {@link RecordingStreamSupport} which delegates to an underlying
+     * `jdk.jfr.consumer.RecordingStream` if the runtime supports it, or {@link Optional#empty()} otherwise.
+     */
+    public static Optional<RecordingStreamSupport> newRecordingStream(Configuration configuration) {
+        return isSupported().map(_v -> {
+            try {
+                return new ReflectiveRecordingStreamSupport(configuration);
+            } catch (Throwable t) {
+                log.warn("JFR recording stream support is not available", t);
+                return null;
+            }
+        });
+    }
+
+    public interface RecordingStreamSupport extends AutoCloseable {
```

Comment:
I think we should declare `@Override void close()` to narrow `close` from `AutoCloseable` not to throw a checked exception, matching `RecordingStream.close`.
This makes it easier to do something like the follwoing, since you can't throw checked exceptions within a functional method:
```java
RecordingStreams.newRecordingStream()
     .ifPresent(recordingStream -> {
         try (recordingStream) {
             ///etc
         }
     });

Reply:
Looks like the same applies to EventStream as well

Reply:
done